library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,57 @@
+[JLBP-4] Avoid dependencies on unstable libraries and features
+--------------------------------------------------------------
+
+- At a high level, don't depend on things that can have frequent breaking
```

Comment:
Oh, and separately: It's probably worth distinguishing the projects that make "breaking changes" only when they change Java packages and artifacts from the projects that make _true_ breaking changes, since the latter are more dangerous. I know that there's going to be a whole item about this from the perspective of the library owner, but it could be useful to discuss here from the perspective of the library user.

Reply:
True, but arguably, it's not great to depend on something that creates a new package with every breaking change, and they do so every six months. That is littering the namespace with packages and bloat. Additionally, such a library will be less likely to continue supporting the old major versions than a library that doesn't do that. Do you think there is a different choice library users should make depending on this distinction? We'd probably have to consider all combinations:

* frequent changes, no renaming (old Guava)
* frequent changes, with renaming (I don't personally have examples)
* infrequent changes, no renaming (new Guava)
* infrequent changes, with renaming (ok http)

My guess is that it's just the frequency dimension that matters (but I'm interested to hear if you disagree).

Reply:
Seeing the comment by @briandealwis , and thinking about frequency: Maybe the right thing to say is "Don't depend on something that can make breaking changes more frequently than the cadence you want for your own library." That covers the core reason for the rule, and leaves open the exact frequency to be elaborated in another best practice (candidate: JLBP-10).

Reply:
Good point about the matrix of possibilities.

I do think that renaming matters in at least some cases. As in our other thread, part of the difference may be the Guava perspective vs. the Google Cloud Library perspective. If Guava had to depend on an unstable HTTP library, I'd much prefer that it be the with-renaming OkHttp model. We might not want to _switch_ to the new OkHttp every time it comes out (especially if releases were frequent, though as you point out, that's not the case for OkHttp), but we'd like the idea that we can keep using OkHttp 1.0 forever, even as the rest of the world moves onward to OkHttp 42.0. Or we can move to OkHttp 42.0 if that's better for the ecosystem. But at least we have a choice. The package bloat is sad but I still think less sad than package conflicts.

Also: You noted that "frequent changes, with renaming" seems rare. If "renaming" typically implies "infrequent changes," then, practically speaking, I think we can say good things about renaming, and users will take the action that we want. If someday OkHttp does decide to release version 42.0, we can revisit :)

Hmm, you know, new line of thinking: I'm probably partially just uncomfortable with referring to what OkHttp does as "breaking changes" at all. I mean, it _is_ in the sense that, if you remove the OkHttp 1 jar from your classpath and add the OkHttp 2 jar, your code breaks :) But it would be great to have some terminology to distinguish this from what I've called "true" breaking changes (perhaps ironically, since "true" breaking changes might or might not break users, but package renames definitely do!). I'm not sure what to call them... "in-place breaking changes" vs. something? For purposes of this doc, maybe we can just say "dependencies that release incompatible versions?" But arguably OkHttp1 and OkHttp2 are "incompatible" in that the types don't interoperate....